Documentation/Resilience/policy_for_type.md

```diff
@@ -0,0 +1,75 @@
+---
+title: Policy for type
+description: Overview of how to work with a policy for a type
+keywords: Overview, Resilience
+author: einari
+weight: 3
+---
+Policies for specific types are accessible and defined linked to a type. These can
+be very useful to use for specific resources needing resilience and makes it easy
+and clear to scope it to that resource specifically.
+
+{{% notice information %}}
+You don't have to define a policy for a type, if someone tries to access an undefined
+policy - the [default policy]({{< relref default_policy >}}) will be used.
+{{% /notice %}}
+
+## Defining a policy for a type
+
+To define a well known policy, all you need to do is add a C# class to your
+project that implements the `ICanDefinePolicyForType` interface.
+From this, you leverage the API of [The Polly Project](https://github.com/App-vNext/Polly/wiki)
+and return the actual policy or sets of policies that will be used.
+
+Below is an example of something that will add a default policy:
+
+```csharp
+using Dolittle.Resilience;
+
+public class NamedPolicyDefiner : ICanDefinePolicyForType
+{
+    public Type Name => typeof(MyResource);
+
+    public Polly.Policy Define()
+    {
+        return Polly.Policy.NoOps();
+    }
+}
+```
+
+{{% notice warning %}}
+There can only be one definition of a policy per type. An exception will be
+thrown if you try to add multiple definitions.
+{{% /notice %}}
+
+## Usage
+
+To make use of a policy for a type, all you need to do is take a dependency to
+`IPolicyFor<MyResource>` in your constructor. This will then automatically resolve
+to the correct policy. Then, the code that needs resilience around it simply calls
+the policy's correct `Execute` method.
+
+Below is an example of how to use it:
+
+```csharp
+using Dolittle.Resilience;
+
+public class MyResource
+{
+    IPolicyFor<MyResource> _policy;
+
+    public MyResource(IPolicyFor<MyResource> policy)
```

Comment:
Would it be possible to deduce the type that is requesting the policy, and then use that to fetch it instead of the type parameter? IMHO that would result in a cleaner interface - more in line with IoC.

Right now it seems like the "common" way to use this would be to always request an `IPolicyFor<MySelf>`, because if you ask for an `IPolicy` you would always get the default one, without the any way of changing. Or am I messing up the Open-Closed principle now...?

Reply:
See your point. We don't have the information readily available right now to what is requesting it (taking a dependency to it). That would involve adding that information to the new `BindingContext` object and get it (if we can) from Autofac, and also retrofitting this into the BootContainer for consistency. 

I agree with you in principle, but I think it is a bit more involved. With this we could also be having a policy for a well known type and share that between two different systems, it is not necessarily the system that is taking the dependency that owns the resource or actually represent the resource. For instance, when we have `IMongoCollection` and we want something for resilience for it, we would maybe have repository wrapper around it and we'd take a dependency to `IPolicyFor<IMongoCollection>` in the repository wrapper (e.g. `IReadModelRepositoryFor<>`). 

Suggest we keep this as-is right now, and something we could potentially optimize for later - if we expand on the things mentioned.

Reply:
Ahh, ok. Then it's definitely more involved. For the future; I was not suggesting to remove the ability to get the policy for any type with the `IPolicyFor<...>`, as you say that makes sense elsewhere, just to change the default behaviour.

But then let's leave that for later 👍